bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/Utils.java

```diff
@@ -107,4 +130,113 @@ public static String generateId(String prefix, String path) {
         return StringUtils.join(prefix, ID_SEPARATOR, StringUtils.substring(DigestUtils.sha256Hex(path), 0, 10));
     }
 
+    /**
+     * Returns a set of resource types for components used to render a given page, including those
+     * from the page template and embedded experience templates.
+     *
+     * @param page the {@link Page}
+     * @param request the current request
+     * @param modelFactory the {@link ModelFactory}
+     * @return
+     */
+    @NotNull
+    public static Set<String> getPageResourceTypes(@NotNull Page page, @NotNull SlingHttpServletRequest request, @NotNull ModelFactory modelFactory) {
+        Set<String> resourceTypes = new HashSet<>();
+        resourceTypes.addAll(getResourceTypes(page.getContentResource(), request, modelFactory));
+        resourceTypes.addAll(getTemplateResourceTypes(page, request, modelFactory));
+        return resourceTypes;
+    }
+
+    /**
+     * Returns a set of resource types for components used to render a given resource,
+     * including it's direct children
+     *
+     * @param resource the resource
+     * @param request the current request
+     * @param modelFactory the {@link ModelFactory}
+     *
+     * @return a set of resource types for components used to render the resource
+     */
+    @NotNull
+    public static Set<String> getResourceTypes(@NotNull Resource resource, @NotNull SlingHttpServletRequest request, @NotNull ModelFactory modelFactory) {
+        Set<String> resourceTypes = new HashSet<>();
+        resourceTypes.add(resource.getResourceType());
+        resourceTypes.addAll(getXFResourceTypes(resource, request, modelFactory));
+        for (Resource child : resource.getChildren()) {
+            //TODO: check it's a cq:Component, used to be allowed node (filtered out by regex)
```

Comment:
A check for `cq:Component` is not needed as we check all the descendant resources and get their resource types. 
I'd remove this comment.

Reply:
Done in #1100